### ISSUE TITLE
GenericAlias' attributes seems to be missing (py 3.12)

### ISSUE THREAD

**Issue by yoann9344** [python/typeshed#12098]
There is no `__or__` in `GenericAlias` defined in [_stdlib/types.pyi_](https://github.com/python/typeshed/blob/6220c20d9360b12e2287511587825217eec3e5b5/stdlib/types.pyi#L577-L596) :
```python
    class GenericAlias:
        @property
        def __origin__(self) -> type: ...
        @property
        def __args__(self) -> tuple[Any, ...]: ...
        @property
        def __parameters__(self) -> tuple[Any, ...]: ...
        def __new__(cls, origin: type, args: Any) -> Self: ...
        def __getitem__(self, typeargs: Any, /) -> GenericAlias: ...
        def __eq__(self, value: object, /) -> bool: ...
        def __hash__(self) -> int: ...
        if sys.version_info >= (3, 11):
            @property
            def __unpacked__(self) -> bool: ...
            @property
            def __typing_unpacked_tuple_args__(self) -> tuple[Any, ...] | None: ...

        # GenericAlias delegates attr access to `__origin__`
        def __getattr__(self, name: str) -> Any: ...
```
As written, it should delegates through `__getattr__`, but :
```python
t = type[str, list]

print(object.__getattribute__(t, "__or__"))
# <method-wrapper '__or__' of types.GenericAlias object at 0x7b48cdb09da0>
print(t.__or__)
# <slot wrapper '__or__' of 'type' objects>
```
The method `__getattribute__` should raise an `AttributeError` if `__or__` would not exist in `GenericAlias` instances. You can read Python's doc about [descriptor](https://docs.python.org/3/howto/descriptor.html#invocation-from-an-instance) :
<details>
<summary>You can reproduce the « delegated » behaviour this way.</summary>

```python
class Plop:
    origin: type = type
    def plop(self): ...

    def __getattr__(self, attr: str):
        return getattr(self.__origin__, attr)

p = Plop()

print(p.plop)
print(p.__or__)
print(object.__getattribute__(p, "plop"))
print(object.__getattribute__(p, "__or__"))
```
</details>

What attribute might also be missing ?
```python
for attr in sorted(set(dir(type[...])) - set(dir(object))):
    with contextlib.suppress(AttributeError):
        object.__getattribute__(t, attr)
        print(attr)

```
<details>
<summary>Stdout</summary>

```python
__args__
__call__
__instancecheck__
__mro_entries__
__or__
__origin__
__parameters__
__ror__
__subclasscheck__
__typing_unpacked_tuple_args__
__unpacked__
```
</details>

**Comment by AlexWaygood** [comment]
We deliberately don't have `GenericAlias.__call__` in the stub, since it would be really difficult to get the signature right, and might mess up the special casing that type checkers already do for `GenericAlias`:

https://github.com/python/typeshed/blob/6220c20d9360b12e2287511587825217eec3e5b5/stdlib/%40tests/stubtest_allowlists/py310.txt#L54

I also don't seem uch point in adding `__mro_entries__`, `__subclasscheck__` or `__instancecheck__` -- these are really implementation details at runtime; they're sort-of useless for type checkers. But I think you're correct that `__or__` and `__ror__` should probably be added! These methods are added specially in CPython (https://github.com/python/cpython/blob/e83ce850f433fd8bbf8ff4e8d7649b942639db31/Objects/genericaliasobject.c#L968), and `__getattr__` wouldn't suffice here, since special methods are looked up on the class rather than the instance.

PR welcome!

**Comment by yoann9344** [comment]
@AlexWaygood thx for the details
`__or__` and `__ror__` seem to be implemented since 3.10 :
```bash
python3.10 /tmp/tmp9xtmyp9s.py
# <method-wrapper '__or__' of types.GenericAlias object at 0x70312b8f2340>
python3.9 /tmp/tmp9xtmyp9s.py
# Traceback (most recent call last):
#   File "/tmp/tmp9xtmyp9s.py", line 12, in <module>
#     print(object.__getattribute__(t, "__or__"))
# AttributeError: 'types.GenericAlias' object has no attribute '__or__'           
```

So it should be like this :
```diff
 if sys.version_info >= (3, 9):
     class GenericAlias:
         @property
         def __origin__(self) -> type: ...
         @property
         def __args__(self) -> tuple[Any, ...]: ...
         @property
         def __parameters__(self) -> tuple[Any, ...]: ...
         def __new__(cls, origin: type, args: Any) -> Self: ...
         def __getitem__(self, typeargs: Any, /) -> GenericAlias: ...
         def __eq__(self, value: object, /) -> bool: ...
         def __hash__(self) -> int: ...
+        if sys.version_info >= (3, 10):
+            def __or__(self, value: Any, /) -> types.UnionType: ...
+            def __ror__(self, value: Any, /) -> types.UnionType: ...
         if sys.version_info >= (3, 11):
             @property
             def __unpacked__(self) -> bool: ...
             @property
             def __typing_unpacked_tuple_args__(self) -> tuple[Any, ...] | None: ...
 
         # GenericAlias delegates attr access to `__origin__`
         def __getattr__(self, name: str) -> Any: ...

```

**Comment by AlexWaygood** [comment]
@yoann9344 yup!
